Add Pylontech integration

danielschramm opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->

Added support for Pylontech US battery packs which are often used in solar systems.
Reading data is currently only supported via locally attached RS485 <-> USB interface.

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

There are a lot of people using these batteries.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: none
- This PR is related to issue: none
- Link to documentation pull request: [#23392](https://github.com/home-assistant/home-assistant.io/pull/23392)


## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

Currently no extra explicit test for this integration is added. If needed please point out in which integration a good example is included as I'm mainly Embedded C++ Dev and still need to learn good Python style from perfect examples.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

danielschramm committed:
Simulated data

just some Stack parameters with simulated data

danielschramm committed:
Simulated data

just some Stack parameters with simulated data

danielschramm committed:
read from dev

danielschramm committed:
read from dev

danielschramm committed:
Communication successful

danielschramm committed:
Communication successful

danielschramm committed:
Use PyPi Package

danielschramm committed:
Use PyPi Package

danielschramm committed:
Merge branch 'home-assistant:dev' into pylontech

danielschramm committed:
Merge branch 'pylontech' of https://github.com/danielschramm/core into pylontech

danielschramm committed:
Merge branch 'home-assistant:dev' into pylontech

danielschramm committed:
Entities Enabled

danielschramm committed:
Entities Enabled

danielschramm committed:
Sensor Factory added

danielschramm committed:
Sensor Factory added

danielschramm committed:
Enable Factory

danielschramm committed:
Enable Factory

danielschramm committed:
entry_id bugfix

danielschramm committed:
entry_id bugfix

danielschramm committed:
Merge branch 'home-assistant:dev' into pylontech

danielschramm committed:
consistant naming

danielschramm committed:
consistant naming

homeassistant commented:
Hi @danielschramm, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#74893).

Once you do that we will be able to review and accept this pull request.

Thanks!

danielschramm committed:
Merge branch 'dev' into pylontech

danielschramm committed:
Merge branch 'dev' into pylontech

danielschramm committed:
Merge branch 'dev' into pylontech

danielschramm committed:
Merge branch 'pylontech' of https://github.com/danielschramm/core into pylontech

danielschramm committed:
Merge branch 'home-assistant:dev' into pylontech

danielschramm committed:
Lib update

Lib Update allows to use network URLs for RS486 to network adapters besides local usb adapters.
No Codechange to Integration needed. Doc update might be useful.

danielschramm committed:
Lib update

Lib Update allows to use network URLs for RS486 to network adapters besides local usb adapters.
No Codechange to Integration needed. Doc update might be useful.

danielschramm committed:
Merge branch 'dev' into pylontech

danielschramm committed:
Merge branch 'home-assistant:dev' into pylontech

danielschramm committed:
Merge branch 'home-assistant:dev' into pylontech

allenporter commented on homeassistant/components/pylontech_us/__init__.py:
```diff
@@ -0,0 +1,96 @@
+        async with async_timeout.timeout(15):
+            await self._hass.async_add_executor_job(self.update)
+            return self._stack.pylonData
+        # except TypeError as err:
```
Don't check in commented out code (here and elsewhere)

allenporter commented on homeassistant/components/pylontech_us/__init__.py:
```diff
@@ -0,0 +1,96 @@
+            device=port, baud=baud, manualBattcountLimit=battery_count
+        )
+        self._last_result = self._stack.update()
+        self._hass = hass
```
This is already set as `self.hass` so this `_hass` can be removed.

danielschramm replied:
Done.

allenporter commented on homeassistant/components/pylontech_us/__init__.py:
```diff
@@ -0,0 +1,96 @@
+        self._stack = PylontechStack(
+            device=port, baud=baud, manualBattcountLimit=battery_count
+        )
+        self._last_result = self._stack.update()
```
Presumably this does blocking I/O here and should not run in the constructor. If the idea is to get a value right away the i think `async_config_entry_first_refresh` is what you may want.

danielschramm replied:
Thanks for the hint. - Moved.

allenporter commented on homeassistant/components/pylontech_us/config_flow.py:
```diff
@@ -0,0 +1,119 @@
+        )
+        stack.update()
+
+        if stack is None:
```
It `stack` is `None` then would the `update()` method would have failed above?

danielschramm replied:
Rewrite of pylontech lib to pass exception up to homeassistant in case of failing connection and update.

allenporter commented on homeassistant/components/pylontech_us/config_flow.py:
```diff
@@ -0,0 +1,119 @@
+            self.hub.validate_config_input
+        )  # !!! no Brackets !!!
+
+        return_data = data
+        return_data["title"] = "Pylontech"
```
Is this ever used? The return value of this function is ignored also.

danielschramm replied:
Removed return values. Exception propagats connect errors.

allenporter reviewed (changes requested):
Hi, I started reviewing this but then it seemed like it doesn't work yet given all the commented out code and the sensor functions are not implemented. Is that right? If so, want to re-open this PR when you'd it's a little further along and ready for review?

danielschramm commented:
> Hi, I started reviewing this but then it seemed like it doesn't work yet given all the commented out code and the sensor functions are not implemented. Is that right? If so, want to re-open this PR when you'd it's a little further along and ready for review?

@allenporter thanks for all your hints.
I'm new to writing integrations so it might take some time, getting everything right. Perhaps you could point me one lightweight integration where I could read a perfect code as reference.
Thanks, Daniel Schramm

allenporter commented:
> > Hi, I started reviewing this but then it seemed like it doesn't work yet given all the commented out code and the sensor functions are not implemented. Is that right? If so, want to re-open this PR when you'd it's a little further along and ready for review?
> 
> @allenporter thanks for all your hints. I'm new to writing integrations so it might take some time, getting everything right. Perhaps you could point me one lightweight integration where I could read a perfect code as reference. Thanks, Daniel Schramm

I just reviewed NextDNS and thought it looked pretty good, but i didn't look to see how close it is to what you're doing.

danielschramm committed:
Merge branch 'home-assistant:dev' into pylontech

danielschramm committed:
Minor cleanup and LINT Errors.

danielschramm committed:
Minor cleanup and LINT Errors.

danielschramm committed:
Remove comments

danielschramm committed:
Remove comments

balloob commented:
If this is ready for review, please rebase the branch and mark it ready. Otherwise let's close it.

danielschramm commented:
Sorry for this beeing so long open.
I hope to get it finished over the Christmas holidays.

danielschramm committed:
Merge branch 'pylontech' of https://github.com/danielschramm/core into pylontech

danielschramm committed:
Merge branch 'home-assistant:dev' into pylontech

danielschramm committed:
Merged files

danielschramm committed:
Merge remote-tracking branch 'origin/dev' into pylontech

danielschramm committed:
cleanup check function.

danielschramm committed:
Merge remote-tracking branch 'origin/dev' into pylontech

danielschramm committed:
Coordinator to other file, better idle handling.

danielschramm committed:
Merge branch 'home-assistant:dev' into pylontech

danielschramm committed:
Merge branch 'home-assistant:dev' into pylontech

danielschramm committed:
Changed kW to W

danielschramm commented:
Project is not stalled.
Due to very low solar radiation in the last nearly Month I had issues getting the Batteries on for a longer time.
I did invest work testing the behavior for shutdown times but need to do more testing on battery on then the weather is better.
Also pulled in the changes from recent development tree.
Updated release of the underlying Python lib is requested to be published in a pull request.

mschiller commented:
I'm super excited! What's the current state?

danielschramm commented:
@mschiller I'm currently stuck a bit - Had time over christmas but could only test and improve Timeout behaviors since we had no sunny days and battery was always off. Now again have to rebase the code and continue development.

If you are interested you might help on testing - having the underlying python library tested against more different battery stacks would be nice. (https://github.com/Tomcat42/pylontech-python) - It needs to be really stable and have a new release to PIP before we can merge this pull request.

danielschramm commented:
Sorry but I will cancel this one as it takes too much time to get communication backend and the plugin integrated into different projects with different requirements.
Probably I will come up with something for MQTT which can work with homeassistant like I already did for other equipment.
